GradeServer.c

```diff
@@ -1,146 +1,216 @@
+    server.sin_family = AF_INET;
+    server.sin_addr.s_addr = INADDR_ANY;
+    server.sin_port = htons(port_number);
+
+    // Bind
+    if (bind(socket_desc, (struct sockaddr *)&server, sizeof(server)) < 0)
+    {
+        // print the error message
+        perror("bind failed. Error");
         return 1;
     }
+    puts("bind done");
+
+    // Listen
+    listen(socket_desc, 3);
+
+    for (int i = 0; i < NUM_THREADS; i++)
+        pthread_create(&threads[i], NULL, process_tasks, NULL);
+
+    // Accept and incoming connection
+    puts("Waiting for incoming connections...");
+    c = sizeof(struct sockaddr_in);
+    // accept connection from an incoming client
+    client_sock = accept(socket_desc, (struct sockaddr *)&client, (socklen_t *)&c);
+    if (client_sock < 0)
+    {
+        perror("accept failed");
+        return 1;
+    }
+    puts("Connection accepted");
+    
+    // Receive a message from client and add it to the task list
+    while (1)
+    {
+      while ((read_size = recv(client_sock, client_message, 2000, 0)) > 0)
+      {
+        // Add the message to the task list
+        task_t task;
+        strcpy(task.message, client_message);
+        add_task(task);
+        // Clear the client_message buffer
+        memset(client_message, 0, 2000);
+        read_size = 0;
+      }
+      if (read_size == -1)
+      {
+        perror("recv failed");
+      }
+    }
+    for (int i = 0; i < NUM_THREADS; i++)
+      pthread_join(threads[i], NULL);
 
-    int num_users = 0;  // number of users read from the file
+    return 0;
+}
 
-    // Read each line of the file
-    char line[MAX_LINE_LENGTH];
-    while (fgets(line, MAX_LINE_LENGTH, file) != NULL) {
 
-        char* id = strtok(line, ":");
-        char* password = strtok(NULL, ":");
 
-        // Store the ID and password in the data structure
-        strcpy(users[num_users].id, id);
-        strcpy(users[num_users].password, password);
-        users[num_users].userType = user_type;
-                if (user_type == STUDENT_USER_TYPE) {
-            users[num_users].grade = 0;
-        }
-        num_users++;
-    }
+void sigint_handler(int signum)
+{
+    printf("Received SIGINT\n");
+    exit(1);
+}
+
+void add_task(task_t task)
+{
+    pthread_mutex_lock(&task_mutex);
 
-    fclose(file);
-    return num_users;
+    // Add the task to the list
+    tasks[num_tasks++] = task;
+
+    pthread_cond_signal(&task_cond);
+    pthread_mutex_unlock(&task_mutex);
 }
 
-void add_task_to_queue(int connfd) {
-    // Add the task (connection) to the task list
-    pthread_mutex_lock(&task_list_mutex);
-    task_list[task_list_size++] = connfd;
-    pthread_cond_signal(&task_list_cond);
-    pthread_mutex_unlock(&task_list_mutex);
+task_t get_task(void)
+{
+    pthread_mutex_lock(&task_mutex);
+
+    while (num_tasks == 0)
+        pthread_cond_wait(&task_cond, &task_mutex);
+
+    // Get the next task from the list
+    task_t task = tasks[--num_tasks];
```

Comment:
We reduce the num of tasks both in get + delete tasks. Wouldn't it create a bug?

Reply:
You're right, we need to remove "delete_task". I missed it.